src/utilities/graphql/transform.ts

```diff
@@ -86,131 +85,330 @@ function nullIfDocIsEmpty(doc: DocumentNode) {
 function getDirectiveMatcher(
   directives: (RemoveDirectiveConfig | GetDirectiveConfig)[],
 ) {
-  return function directiveMatcher(directive: DirectiveNode) {
-    return directives.some(
-      dir =>
-        (dir.name && dir.name === directive.name.value) ||
-        (dir.test && dir.test(directive)),
-    );
+  const nameSet = new Set<string>();
+  const tests: Array<(directive: DirectiveNode) => boolean> = [];
+  directives.forEach(directive => {
+    if (directive.name) {
+      nameSet.add(directive.name);
+    } else if (directive.test) {
+      tests.push(directive.test);
+    }
+  });
+
+  return (directive: DirectiveNode) => (
+    nameSet.has(directive.name.value) ||
+    tests.some(test => test(directive))
+  );
+}
+
+// Helper interface and function used by removeDirectivesFromDocument to keep
+// track of variable references and fragments spreads found within a given
+// operation or fragment definition.
+interface InternalInUseInfo {
+  variables: Set<string>;
+  fragmentSpreads: Set<string>;
+  // Set to true when we deliberately remove a fragment definition, so we can
+  // make sure also to remove dangling ...spreads that refer to it.
+  removed?: boolean;
+  // Populated by the populateTransitiveVars helper function below.
+  transitiveVars?: Set<string>;
+}
+function makeInUseGetterFunction<TKey>() {
+  const map = new Map<TKey, InternalInUseInfo>();
+
+  return function getInUse(key: TKey): InternalInUseInfo {
```

Comment:
Nit: there are two `getInUse` functions in different closures a few lines apart.

```suggestion
  return function inUseGetterFunction(key: TKey): InternalInUseInfo {
```